src/lang/base/ADTChecker.ml

```diff
@@ -284,16 +284,20 @@ module ScillaRecursion
 
 
   let recursion_rprins_elibs recursion_principles ext_libs libs =
-    let ((rec_elibs, adts), emsgs) =
-      List.fold_left ext_libs ~init:(([], []), [])
+
+    let rec recurser libl =
+      List.fold_left libl ~init:(([], []), [])
         ~f:(fun ((rec_elibs_acc, adts_acc), emsgs_acc) ext_lib ->
             let ((rec_elib, elib_adt), emsg) =
-              match recursion_library ext_lib with
-              | Ok (lib, adt) -> ((lib :: rec_elibs_acc, adt :: adts_acc), emsgs_acc)
-              | Error el -> ((rec_elibs_acc, adts_acc), emsgs_acc @ el) in
+              let ((rec_dep_libs, dep_adt), dep_emsgs) = recurser ext_lib.deps in
+              match recursion_library ext_lib.libn with
+              | Ok (lib, adt) ->
+                let (libn' : RecursionSyntax.libtree) = { libn = lib; deps = rec_dep_libs } in
+                ((rec_elibs_acc @ [libn'], adts_acc @ dep_adt @ adt), emsgs_acc @ dep_emsgs)
```

Comment:
I think it can be, but the reversing has to be done at each level of the tree (not just at once as we did before). Since the expected length of these lists is small, I figured that this may be more straightforward.

Reply:
OK, no problem.